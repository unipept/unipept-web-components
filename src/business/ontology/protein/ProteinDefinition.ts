import Definition from "./../Definition";
import { NcbiId } from "./../taxonomic/ncbi/NcbiTaxon";
import { EcCode } from "./../functional/ec/EcDefinition";
import { GoCode } from "./../functional/go/GoDefinition";
import { InterproCode } from "./../functional/interpro/InterproDefinition";

export type UniprotAccessionId = string;

export default class ProteinDefinition implements Definition {
    constructor(
        public readonly uniprotAccessionId: UniprotAccessionId,
        public readonly name: string,
        public readonly organism: NcbiId,
        public readonly ecNumbers: EcCode[],
        public readonly goTerms: GoCode[],
        public readonly interproEntries: InterproCode[]
    ) {}
}
