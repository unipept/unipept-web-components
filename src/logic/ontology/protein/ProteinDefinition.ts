import Definition from "../Definition";
import { EcCode } from "../functional";
import { GoCode } from "../functional";
import { InterproCode } from "../functional";
import { NcbiId } from "../taxonomic";
import UniprotAccessionId from "./UniprotAccessionId";

export default class ProteinDefinition implements Definition {
    constructor(
        public readonly uniprotAccessionId: UniprotAccessionId,
        public readonly name: string,
        public readonly organism: NcbiId,
        public readonly ecNumbers: EcCode[],
        public readonly goTerms: GoCode[],
        public readonly interproEntries: InterproCode[]
    ) {}
}
