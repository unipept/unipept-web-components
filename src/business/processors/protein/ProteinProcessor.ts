import { Peptide } from "./../../ontology/raw/Peptide";
import ProteinDefinition from "./../../ontology/protein/ProteinDefinition";
import ProteinResponseCommunicator from "./../../communication/protein/ProteinResponseCommunicator";

export default class ProteinProcessor {
    /**
     * Returns a list of protein definitions associated with the given peptide. These definitions are guaranteed to be
     * sorted by organism name.
     *
     * @param peptide The peptide for which we need to retrieve all protein definitions.
     * @param equateIl Whether the amino acids "I" and "L" in a peptide should be treated equally.
     */
    public async getProteinsByPeptide(peptide: Peptide, equateIl: boolean): Promise<ProteinDefinition[]> {
        const proteinResponseCommunicator = new ProteinResponseCommunicator();
        const responses = await proteinResponseCommunicator.getResponse(peptide, equateIl);

        return responses.map(response => new ProteinDefinition(
            response.uniprotAccessionId,
            response.name,
            response.organism,
            response.ecNumbers,
            response.goTerms,
            response.interproEntries
        ));
    }
}
