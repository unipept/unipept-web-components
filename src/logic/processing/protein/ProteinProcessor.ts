import { ProteinResponseCommunicator } from "@/logic/communication/protein";
import { NcbiId, Peptide } from "@/logic/ontology";
import { ProteinDefinition } from "@/logic/ontology/protein";

export default class ProteinProcessor {
    private proteins: ProteinDefinition[];
    private lca: NcbiId | undefined;
    private commonLineage: NcbiId[];

    constructor() {
        this.proteins = [];
        this.commonLineage = [];
    }

    /**
     * Returns a list of protein definitions associated with the given peptide. These definitions are guaranteed to be
     * sorted by organism name.
     */
    public getProteins(): ProteinDefinition[] {
        return this.proteins;
    }

    /**
     * Returns the id of the lowest common ancestor NCBI-taxon for the given peptide and search configuration. Returns
     * undefined in the event that no LCA was found.
     */
    public getLca(): NcbiId | undefined {
        return this.lca
    }

    /**
     * Compute the common lineage for all organisms associated with the proteins that matched with the given peptide.
     */
    public getCommonLineage(): NcbiId[] {
        return this.commonLineage;
    }

    /**
     * @param peptide The peptide for which we need to retrieve all protein definitions.
     * @param equateIl Whether the amino acids "I" and "L" in a peptide should be treated equally.
     */
    public async compute(peptide: Peptide, equateIl: boolean): Promise<void> {
        const proteinResponseCommunicator = new ProteinResponseCommunicator();
        const response = await proteinResponseCommunicator.getResponse(peptide, equateIl);

        if (response) {
            this.proteins = response.proteins.map(response => new ProteinDefinition(
                response.uniprotAccessionId,
                response.name,
                response.organism,
                response.ecNumbers,
                response.goTerms,
                response.interproEntries
            ));

            this.lca = response.lca;
            this.commonLineage = response.common_lineage;
        }
    }
}