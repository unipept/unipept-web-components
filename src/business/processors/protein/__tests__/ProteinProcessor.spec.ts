import Setup from "@/test/Setup";
import ProteinProcessor from "@/business/processors/protein/ProteinProcessor";
import ProteinDefinition from "@/business/ontology/protein/ProteinDefinition";

describe("ProteinProcessor", () => {
    beforeAll(() => {
        const setup = new Setup();
        setup.setupAll();
    })

    it("correctly parses protein responses", async() => {
        const proteinProcessor = new ProteinProcessor();
        const proteinDefinition: ProteinDefinition[] = await proteinProcessor.getProteinsByPeptide("AALTER", true);

        expect(proteinDefinition.length).toEqual(8);
        expect(proteinDefinition).toMatchSnapshot();
    })
});
