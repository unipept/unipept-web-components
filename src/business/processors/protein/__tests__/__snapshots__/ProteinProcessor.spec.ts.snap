// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProteinProcessor correctly parses protein responses 1`] = `
Array [
  ProteinDefinition {
    "ecNumbers": Array [
      "2.3.2.27",
    ],
    "goTerms": Array [
      "GO:0005737",
      "GO:0005634",
      "GO:0070696",
      "GO:0004842",
      "GO:0016567",
    ],
    "interproEntries": Array [
      "IPR011989",
      "IPR016024",
      "IPR000225",
      "IPR003613",
      "IPR013083",
    ],
    "name": "U-box domain-containing protein 45",
    "organism": 3702,
    "uniprotAccessionId": "Q9C7G1",
  },
  ProteinDefinition {
    "ecNumbers": Array [
      "5.6.2.1",
      "3.6.4.12",
    ],
    "goTerms": Array [
      "GO:0005737",
      "GO:0005524",
      "GO:0003677",
      "GO:0003678",
      "GO:0003917",
      "GO:0046872",
      "GO:0000746",
      "GO:0008152",
    ],
    "interproEntries": Array [
      "IPR014059",
      "IPR014129",
      "IPR009767",
      "IPR027417",
      "IPR040668",
      "IPR040987",
      "IPR014862",
    ],
    "name": "DNA helicase I",
    "organism": 83333,
    "uniprotAccessionId": "P14565",
  },
  ProteinDefinition {
    "ecNumbers": Array [],
    "goTerms": Array [
      "GO:0005884",
      "GO:0005737",
      "GO:0005789",
      "GO:0005634",
      "GO:0008017",
      "GO:0045010",
      "GO:0007275",
    ],
    "interproEntries": Array [
      "IPR015425",
      "IPR042201",
      "IPR001265",
    ],
    "name": "Formin",
    "organism": 9031,
    "uniprotAccessionId": "Q05858",
  },
  ProteinDefinition {
    "ecNumbers": Array [
      "3.1.3.3",
    ],
    "goTerms": Array [
      "GO:0005737",
      "GO:0005829",
      "GO:0043005",
      "GO:0005509",
      "GO:0000287",
      "GO:0016791",
      "GO:0004647",
      "GO:0042803",
      "GO:0016311",
      "GO:0001701",
      "GO:0006564",
      "GO:0006563",
      "GO:0009612",
      "GO:0031667",
      "GO:0033574",
    ],
    "interproEntries": Array [
      "IPR036412",
      "IPR023214",
      "IPR023190",
    ],
    "name": "Phosphoserine phosphatase",
    "organism": 9606,
    "uniprotAccessionId": "P78330",
  },
  ProteinDefinition {
    "ecNumbers": Array [
      "2.7.1.50",
    ],
    "goTerms": Array [
      "GO:0005524",
      "GO:0004417",
      "GO:0000287",
      "GO:0009228",
      "GO:0009229",
    ],
    "interproEntries": Array [
      "IPR000417",
      "IPR029056",
    ],
    "name": "Hydroxyethylthiazole kinase",
    "organism": 266940,
    "uniprotAccessionId": "A6W6A1",
  },
  ProteinDefinition {
    "ecNumbers": Array [
      "6.3.2.13",
    ],
    "goTerms": Array [
      "GO:0005737",
      "GO:0005524",
      "GO:0000287",
      "GO:0008765",
      "GO:0007049",
      "GO:0051301",
      "GO:0071555",
      "GO:0009252",
      "GO:0008360",
    ],
    "interproEntries": Array [
      "IPR036565",
      "IPR004101",
      "IPR036615",
      "IPR013221",
      "IPR000713",
      "IPR035911",
      "IPR005761",
    ],
    "name": "UDP-N-acetylmuramoyl-L-alanyl-D-glutamate--2,6-diaminopimelate ligase",
    "organism": 103690,
    "uniprotAccessionId": "Q8YWF0",
  },
  ProteinDefinition {
    "ecNumbers": Array [
      "3.1.3.3",
    ],
    "goTerms": Array [
      "GO:0005509",
      "GO:0000287",
      "GO:0004647",
      "GO:0006564",
      "GO:0006563",
    ],
    "interproEntries": Array [
      "IPR036412",
      "IPR023214",
      "IPR023190",
    ],
    "name": "Phosphoserine phosphatase",
    "organism": 9601,
    "uniprotAccessionId": "Q5RB83",
  },
  ProteinDefinition {
    "ecNumbers": Array [
      "6.3.2.13",
    ],
    "goTerms": Array [
      "GO:0005737",
      "GO:0005524",
      "GO:0000287",
      "GO:0008765",
      "GO:0007049",
      "GO:0051301",
      "GO:0071555",
      "GO:0009252",
      "GO:0008360",
    ],
    "interproEntries": Array [
      "IPR036565",
      "IPR004101",
      "IPR036615",
      "IPR013221",
      "IPR000713",
      "IPR035911",
      "IPR005761",
    ],
    "name": "UDP-N-acetylmuramoyl-L-alanyl-D-glutamate--2,6-diaminopimelate ligase",
    "organism": 240292,
    "uniprotAccessionId": "Q3ME15",
  },
]
`;
